### PR TITLE
connect #630 redirect to returnTo url after logout

### PR DESCRIPTION
**Pull request by DrMegavolt**

### CODE REVIEW

**Comment by DrMegavolt** [comment]
@StickNitro can you please verify it works as you expect

File: lib/policies/oauth2/site.js [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
+const services = require('../../services/index');
 
 module.exports.loginForm = (request, response) => response.render(path.join(__dirname, 'views/login'));
 
 module.exports.login = passport.authenticate('local', { successReturnToOrRedirect: '/', failureRedirect: '/login' });
 
-module.exports.logout = (request, response) => {
-  request.logout();
-  response.redirect('/');
+module.exports.logout = (req, res) => {
```
Comment (XVincentX) [inline lib/policies/oauth2/site.js]: 👍 for the renaming

File: lib/policies/oauth2/site.js [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+  req.logout();
+  if (!req.query.returnTo) {
+    return res.redirect('/'); // TODO: implement settings section for default view path and return redirects
+  }
+  const returnUrl = decodeURI(req.query.returnTo);
```
Comment (XVincentX) [inline lib/policies/oauth2/site.js]: Triple check: I think that express is already decoding the parameters for us under the hood.

File: lib/policies/oauth2/site.js [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
+  if (!req.query.clientId) {
+    return res.redirect(returnUrl);
+  }
+
+  const client = services.application.get(req.query.clientId);
+  if (client && client.redirectUri && client.redirectUri === returnUrl) {
```
Comment (XVincentX) [inline lib/policies/oauth2/site.js]: This is returning a `Promise` that should be resolved before accessing its data. These lines will probably make EG crash.

File: lib/policies/oauth2/site.js [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
+    return res.redirect(returnUrl);
+  }
+
+  res.status(400).send(`Invalid return URI ${returnUrl} for client id ${req.query.clientId}`);
```
Comment (XVincentX) [inline lib/policies/oauth2/site.js]: `req.query.clientId` might be undefined — we should check for it.

File: lib/policies/oauth2/site.js [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
-  request.logout();
-  response.redirect('/');
+module.exports.logout = (req, res) => {
+  req.logout();
+  if (!req.query.returnTo) {
+    return res.redirect('/'); // TODO: implement settings section for default view path and return redirects
+  }
+  const returnUrl = decodeURI(req.query.returnTo);
+  if (!req.query.clientId) {
+    return res.redirect(returnUrl);
```
Comment (XVincentX) [inline lib/policies/oauth2/site.js]: This is a bit concerning to me. 

What's happening here is basically: If you want to redirect to any URL you have, just do not provide the `clientID` and we're good to go.

This seems to be a way to skip the security check — something that usually you do not want to do. I'd not provide this feature at all. No clientId? You get an error, and I won't log you out.

Also — the clientId should be also verified whether it's still active.

File: .dockerignore [hunk 6]
```diff
@@ -1 +1,7 @@
 node_modules/
+.vscode/
+.idea/
+.git/
+.circleci/
```
Comment (XVincentX) [inline .dockerignore]: This is also part of #628 — but that's fine I guess.

File: lib/policies/oauth2/site.js (hunk 5 above)
Comment (DrMegavolt) [reply lib/policies/oauth2/site.js]: replicated from Auth0 behaviour

Comment (XVincentX) [reply lib/policies/oauth2/site.js]: Can you point the documentation link? I just want to check that out.

Comment (DrMegavolt) [reply lib/policies/oauth2/site.js]: see the ticket and https://auth0.com/docs/logout
there is actually a bigger problem here even with this PR it still works incorrectly 
If you have time you can investigate this 

Gitter conversation text ---

Neil Stevens @StickNitro Feb 20 16:04
@DrMegavolt Ok so in full,
Navigate to application (user not logged in so redirects to http://localhost:8080/oauth2/authorize?response_type=token&clientId=...&redirectUrl=...
User is prompted to login (using oauth2 credentials)
User is prompted to authorize application
User is redirected to the application (redirectUrl along with token which is cached in sessionStorage and used to access the backend)
Click logout in application (this clears the persisted token and redirects user to http://localhost:8080/logout`)
Logout url above currently redirects to /
At this point I manually navigate to my application and get Step 3

Serhii Kuts @DrMegavolt Feb 20 16:07
and it should be step 2 ? (assuming step 1 will happen automatically)

Neil Stevens @StickNitro Feb 20 16:07
Checking in Chrome dev tools there is a cookie `connect.sid’ that is still present after logout and redirect to /, if i manually remove this cookie after Step 6 above and manually navigate to my application then I get prompted to login (effectively starting at Step 1)

Comment (XVincentX) [reply lib/policies/oauth2/site.js]: Ok thanks. I'll see if we can prioritize this and look into.

**Comment by kevinswiber** [comment]
@XVincentX @DrMegavolt This has been sitting for 19 days.  Is there a resolution here?

**Comment by XVincentX** [comment]
This is partially my fault, I haven't been really looking into this actively.

It's also missing tests. I can take over and bring this over the finish line, if required.

**Comment by DrMegavolt** [comment]
@XVincentX I cannot work on this at the moment. If it is relevant please take over

**Comment by XVincentX** [comment]
@DrMegavolt I think the feature is worth continuing. I'll prepare a card and discuss it during next sprint planning. Thanks for pinging me on this.

**Pull request closed without merging**